src/Evolution/Particles/MonteCarlo/NeutrinoInteractionTable.cpp

```diff
@@ -0,0 +1,377 @@
+        for (size_t nt = 0; nt < n_points_temperature; nt++) {
+          for (size_t nr = 0; nr < n_points_density; nr++) {
+            idx = ny * n_points_temperature * n_points_density +
+                  nt * n_points_density + nr;
+            gsl::at(gsl::at(table_scattering_opacity, ns), ng)[idx] =
+                buffer[ng][ns][ny][nt][nr] * DistUnit;
+            std::clamp(gsl::at(gsl::at(table_scattering_opacity, ns), ng)[idx],
+                       min_kappa, max_kappa);
+          }
+        }
+      }
+    }
+  }
+
+  // Correction to emissivity/absorption/scattering at end of table to keep
+  // Ye within the table bounds (in this regime, we cannot evolve Ye
+  // according to the true tabulated value; we correct the tabulated
+  // values in the hope of at least getting lepton number conservation).
+  if (NeutrinoSpecies >= 2) {
+    // Correction at bounds of table in electron fraction
+    // assume that we have at least electron neutrinos and
+    // antineutrinos.
+    // We correct the emissivity at both bounds, and the
+    // absorption only when absorption would drive the
+    // electron fraction outside of the table bounds.
+    // Note that absorbing an electron neutrino (ns==0)
+    // increases the electron fraction.
+    for (size_t ng = 0; ng < EnergyBins; ng++) {
+      for (size_t nt = 0; nt < n_points_temperature; nt++) {
+        for (size_t nr = 0; nr < n_points_density; nr++) {
+          idx = nt * n_points_density + nr;
+          gsl::at(gsl::at(table_emissivity, 0), ng)[idx] = 0.0;
+          gsl::at(gsl::at(table_emissivity, 1), ng)[idx] = 0.0;
+          gsl::at(gsl::at(table_absorption_opacity, 1), ng)[idx] = min_kappa;
+
+          idx = (n_points_electron_fraction - 1) * n_points_temperature *
+                    n_points_density +
+                nt * n_points_density + nr;
+          gsl::at(gsl::at(table_emissivity, 0), ng)[idx] = 0.0;
+          gsl::at(gsl::at(table_emissivity, 1), ng)[idx] = 0.0;
+          gsl::at(gsl::at(table_absorption_opacity, 0), ng)[idx] = min_kappa;
+        }
+      }
+    }
+  }
+}
+
+template <size_t EnergyBins, size_t NeutrinoSpecies>
+void NeutrinoInteractionTable<EnergyBins, NeutrinoSpecies>::
+    get_neutrino_matter_interactions(
+        const gsl::not_null<
+            std::array<std::array<DataVector, EnergyBins>, NeutrinoSpecies>*>
+            emissivity_in_cell,
+        const gsl::not_null<
+            std::array<std::array<DataVector, EnergyBins>, NeutrinoSpecies>*>
+            absorption_opacity,
+        const gsl::not_null<
+            std::array<std::array<DataVector, EnergyBins>, NeutrinoSpecies>*>
+            scattering_opacity,
+        const Scalar<DataVector>& electron_fraction,
+        const Scalar<DataVector>& rest_mass_density,
+        const Scalar<DataVector>& temperature,
+        const Scalar<DataVector>& cell_proper_four_volume,
+        const double& minimum_temperature) const {
```

Comment:
Is it possible to use our existing multi-linear interpolation class? `src/NumericalAlgorithms/Interpolation/MultiLinearSpanInterpolation.hpp` This is what the tabulated EOS does, eg line 238 `Tabulated3d.cpp`